lib/pgauge_exchange.cu

```diff
@@ -3,358 +3,258 @@
       x[0] = (2 * idx + xodd)  - za * X[0];
       break;
     }
+
     int id = (((x[3] * X[2] + x[2]) * X[1] + x[1]) * X[0] + x[0]) >> 1;
     typedef complex<Float> Complex;
     typedef typename mapper<Float>::type RegType;
     RegType tmp[NElems];
     Complex data[9];
-    if ( pack ) {
-      arg.dataOr.load(data, id, dir, parity);
+
+    if (pack) {
+      arg.dataOr.load(data, id, arg.dir, arg.parity);
       arg.dataOr.reconstruct.Pack(tmp, data, id);
-      for ( int i = 0; i < NElems / 2; ++i ) array[idx + size * i] = Complex(tmp[2*i+0], tmp[2*i+1]);
-    }
-    else{
+      for ( int i = 0; i < NElems / 2; ++i ) arg.array[idx + arg.size * i] = Complex(tmp[2*i+0], tmp[2*i+1]);
+    } else {
       for ( int i = 0; i < NElems / 2; ++i ) {
-        tmp[2*i+0] = array[idx + size * i].real();
-        tmp[2*i+1] = array[idx + size * i].imag();
+        tmp[2*i+0] = arg.array[idx + arg.size * i].real();
+        tmp[2*i+1] = arg.array[idx + arg.size * i].imag();
       }
-      arg.dataOr.reconstruct.Unpack(data, tmp, id, dir, 0, arg.dataOr.X, arg.dataOr.R);
-      arg.dataOr.save(data, id, dir, parity);
+      arg.dataOr.reconstruct.Unpack(data, tmp, id, arg.dir, 0, arg.dataOr.X, arg.dataOr.R);
+      arg.dataOr.save(data, id, arg.dir, arg.parity);
     }
   }
 
-#ifdef MULTI_GPU
-  static void *send[4];
-  static void *recv[4];
-  static void *sendg[4];
-  static void *recvg[4];
   static void *send_d[4];
   static void *recv_d[4];
   static void *sendg_d[4];
   static void *recvg_d[4];
   static void *hostbuffer_h[4];
   static qudaStream_t GFStream[2];
-  static size_t offset[4];
-  static size_t bytes[4];
-  static size_t faceVolume[4];
-  static size_t faceVolumeCB[4];
-  // do the exchange
   static MsgHandle *mh_recv_back[4];
   static MsgHandle *mh_recv_fwd[4];
   static MsgHandle *mh_send_fwd[4];
   static MsgHandle *mh_send_back[4];
-  static int X[4];
-  static dim3 block[4];
-  static dim3 grid[4];
-  static bool notinitialized = true;
-#endif // MULTI_GPU
+  static int *X;
+  static bool init = false;
 
   /**
    * @brief Release all allocated memory used to exchange data between nodes
    */
-  void PGaugeExchangeFree(){
-#ifdef MULTI_GPU
+  void PGaugeExchangeFree()
+  {
     if ( comm_dim_partitioned(0) || comm_dim_partitioned(1) || comm_dim_partitioned(2) || comm_dim_partitioned(3) ) {
-      cudaStreamDestroy(GFStream[0]);
-      cudaStreamDestroy(GFStream[1]);
-      for ( int d = 0; d < 4; d++ ) {
-        if ( commDimPartitioned(d)) {
-          comm_free(mh_send_fwd[d]);
-          comm_free(mh_send_back[d]);
-          comm_free(mh_recv_back[d]);
-          comm_free(mh_recv_fwd[d]);
-          device_free(send_d[d]);
-          device_free(recv_d[d]);
-          device_free(sendg_d[d]);
-          device_free(recvg_d[d]);
-        #ifndef GPU_COMMS
-          host_free(hostbuffer_h[d]);
-        #endif
+      if (init) {
+        cudaStreamDestroy(GFStream[0]);
```

Comment:
`qudaStreamDestroy`?

Reply:
The streams and events will be dealt with in the next PR